odoo/http.py

```diff
@@ -1447,7 +1448,10 @@ def _dispatch_nodb():
                     return request._handle_exception(e)
                 return result
 
-            with request:
+            with contextlib.ExitStack() as stack:
+                for manager in self.get_dispatch_context_managers(request):
+                    stack.enter_context(manager)
```

Comment:
I'm not sure I like having an extra loop and function call in the main dispatch code path, especially as those will *not* be used almost 100% of the time. Couldn't we construct this in a different way, to reduce the overhead as much as possible in the general case?

maybe something like this, to keep the overhead to 2 extra `if`s when no profiling is active?

```py
ctx_manager = request
sub_ctx_managers = None
if request.session.profile_session:
    ctx_manager = contextlib.ExitStack()
    req_ctx_managers = self.get_dispatch_context_managers(request)

with ctx_manager:
    if sub_ctx_managers:
        for manager in sub_ctx_managers:
            ctx_manager.enter_context(manager)
    # main dispatching
```

Reply:
Another solution would be to add a parameter to Profiler to give another context manager to enter and exit automaticaly

```python
ctx_manager = request
if request.session.profile_session:
    ctx_manager = Profiler(... , ctx_manager=ctx_manager)
with ctx_manager :
    # main dispatching
```
- This would allow to remove the `if self.init_frame.f_code.co_name == 'enter_context':` special case. 
- cleaner in http.py
- could be usefull in other case.

 I didn't choose this solution because it looks far fetched and non standard but if you are afraid of this loop it may be a solution.
 
We can also implement our version of the old 'contextlib.nested' but I'm worried it may break and complexify the detection of the `init_frame`.